### PR TITLE
add multi-query attn

### PR DESCRIPTION
**Pull request by vchiley**
add multi-query attn

test: added to `test_attn_impl`

### CODE REVIEW

**Comment by vchiley** [comment]
With multi-query attention, the key value projection matrix is smaller.
There are fewer Params / FLOPs in the model therefore the throughput is higher, but the MFU is pretty much the same
Screenshot 2023-04-17 at 10 12 54 PM
This is true for attn_impl \in {flash, triton, and torch}
Screenshot 2023-04-17 at 10 14 02 PM

Given there are fewer params, the networks final perf is degraded:
Screenshot 2023-04-17 at 5 00 19 PM
(granted the hparams are set for the multihead attn)


Generally the layer works. I wouldn't recommend but it might be useful in some inference contexts (BW or memory constrained inference settings)

File: examples/llm/src/models/layers/attention.py [hunk 1]
```diff
@@ -347,6 +379,129 @@ def forward(self,
...
+    def __init__(
+        self,
+        d_model: int,
+        n_heads: int,
+        attn_impl: str = 'triton',
+        attn_clip_qkv: Optional[float] = None,
+        attn_qk_ln: bool = False,
+        softmax_scale: Optional[float] = None,
+        attn_pdrop: float = 0.0,
+        low_precision_layernorm: bool = False,
```
Comment (abhi-mosaic) [inline examples/llm/src/models/layers/attention.py:398]: this probably gets refactored after your norm PR goes in? Just heads up

File: examples/llm/src/models/layers/gpt_blocks.py [hunk 2]
```diff
@@ -43,15 +44,17 @@ def __init__(self,
...
                  resid_pdrop: float = 0.0,
                  low_precision_layernorm: bool = False,
+                 multiquery_attention: bool = False,
                  device: Optional[str] = None,
                  **kwargs):
         del kwargs  # unused, just to capture any extra args from the config
         super().__init__()
 
         layernorm_class = LPLayerNorm if low_precision_layernorm else nn.LayerNorm
+        attn_class = MultiQueryAttention if multiquery_attention else MultiheadAttention
```
Comment (abhi-mosaic) [inline examples/llm/src/models/layers/gpt_blocks.py:54]: Do you think we'll stop at 2? Or should we make `attn_class` a choice `str` out of [`multihead`, `multiquery`] the way we do `attn_impl`?

**Review by abhi-mosaic — APPROVED** [review]
Logic reads right to me. Leaning heavily on the WandB plots to say that convergence is happening (albeit a bit degraded). Cool to see the training throughput improvement. Would be very curious to try HF generate on a MQA model and see how much faster it is...

File: examples/llm/src/models/layers/gpt_blocks.py (hunk 2 above)
Comment (vchiley) [reply examples/llm/src/models/layers/gpt_blocks.py:54]: Yes, I wanted to do this as part of this PR, but this'll be a next step, wanted to merge this before the foundry migration.

File: examples/llm/src/models/layers/attention.py (hunk 1 above)
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:398]: yeah merge conflicts make this one pretty clear...

**Pull request merged**